### DISCUSSION TITLE
NER with premade tokens vs string as input.

### DISCUSSION THREAD

**Discussion by dyerrington** [explosion/spaCy#8040] [Help: Coding & Implementations] [answered]
I have a pipeline outputting data from an OCR model that also gives me coordinates for each and every single token.  It's important to maintain that ordered list as the "source of truth" since I have an application that needs to plot bounding boxes for later evaluation.  I've found that I get different results when using `spacy.tokens.doc.Doc` to construct the doc vs unstructured string as the source.  I assume it has something to do with how tokenization is happening but running out of options to investigate based on my experience.

**Entities from a string at the beginning of the pipeline:**
```python
doc = nlp("Purchase Price: Thirty-Three Million Five Hundred Thousand Dollars $33,500,000 for Mr. Potatoehead.")
for ent in doc.ents:
    print(ent.text, ent.start_char, ent.end_char, ent.label_, ent.orth_)
```

**Output:**
```python
Thirty-Three Million 16 36 CARDINAL Thirty-Three Million
33,500,000 68 78 MONEY 33,500,000
Potatoehead 88 99 PERSON Potatoehead
```

**Entities from a token list at the beginning of manually constructed pipeline:**
```python
tokens = ['Purchase', "Price:", "Thirty-Three", "Million", "Five", "Hundred", "Thousand", "Dollars", "$33,500,000", "for", "Mr.", "Potatoehead"]

doc = spacy.tokens.doc.Doc(nlp.vocab, words=tokens, spaces = [True] * len(tokens))
for name, proc in nlp.pipeline:
    doc = proc(doc)
    
for ent in doc.ents:
    print(ent.text, ent.start_char, ent.end_char, ent.label_, ent.orth_)
```

**Output:**
```python
Potatoehead 89 100 PERSON Potatoehead
```

The last idea I looked into was to make sure the steps in the pipeline are the same but I'm sure it has something to do with the tokenizer.  When I inspect the tokens in each `doc`, they look pretty much the same.  I'm also just using the default English core web `nlp = spacy.load("en_core_web_sm")`.

I'm assuming I'm overlooking something very obvious?  Any advice appreciated!  Thanks in advance!

**Comment by dyerrington** [comment, marked as answer]
Looking more into the default tokenizer, it seems that punctuation is essential.  If I break the tokens up by punctuation as well, this ends up producing identical entities:

```
tokens = ['Purchase', "Price", ":", "Thirty", "-","Three", "Million", "Five", "Hundred", "Thousand", "Dollars", "$", "33,500,000", "for", "Mr.", "Potatoehead"]
```

I examined the output of each token structure post init and found that the lemma and pos was affected by punctuation which makes complete sense.

```python
for token in doc:
    print(token.text, token.lemma_, token.pos_, token.tag_, token.dep_,
            token.shape_, token.is_alpha, token.is_stop)
```

![image](https://user-images.githubusercontent.com/2257834/117510261-9b1bf300-af40-11eb-8b93-20386341f04c.png)

I guess what I'm going to do is create a lookup object that can take the original tokens indice that maps to the post-tokenized format of the default tokenizer with punctuation which sounds like a fun problem.

**Reply by adrianeboyd** [reply]
A note: you can provide the right values to `Doc(spaces=)` to get the tokenization that you'd like with the original character offsets.
